pilot/pkg/xds/endpoint_builder.go

```diff
@@ -388,33 +405,51 @@ func buildEnvoyLbEndpoint(proxyless bool, e *model.IstioEndpoint) *endpoint.LbEn
 				Address: addr,
 			},
 		},
+		Metadata: &core.Metadata{},
 	}
 
 	// Istio telemetry depends on the metadata value being set for endpoints in the mesh.
 	// Istio endpoint level tls transport socket configuration depends on this logic
 	// Do not remove pilot/pkg/xds/fake.go
-	ep.Metadata = util.BuildLbEndpointMetadata(e.Network, e.TLSMode, e.WorkloadName, e.Namespace, e.Locality.ClusterID, e.Labels)
+	util.BuildLbEndpointMetadata(e.Network, e.TLSMode, e.WorkloadName, e.Namespace, e.Locality.ClusterID, e.Labels, ep.Metadata)
 
 	address, port := e.Address, e.EndpointPort
 	tunnelAddress, tunnelPort := address, model.HBoneInboundListenPort
 
 	supportsTunnel := false
+	// Other side is a waypoint proxy.
+	if al := e.Labels[constants.ManagedGatewayLabel]; al == constants.ManagedGatewayMeshControllerLabel {
+		supportsTunnel = true
+	}
+	// Otherwise has ambient enabled. Note: this is a synthetic label, not existing in the real Pod.
+	if al := e.Labels[constants.AmbientRedirection]; al == constants.AmbientRedirectionEnabled {
+		supportsTunnel = true
+	}
 	// Otherwise supports tunnel
 	// Currently we only support HTTP tunnel, so just check for that. If we support more, we will
 	// need to pick the right one based on our support overlap.
 	if e.SupportsTunnel(model.TunnelHTTP) {
 		supportsTunnel = true
 	}
-	if proxyless {
+	if b.proxy.IsProxylessGrpc() {
 		// Proxyless client cannot handle tunneling, even if the server can
 		supportsTunnel = false
 	}
-	if !features.EnableHBONE {
+
+	if !b.proxy.EnableHBONE() {
 		supportsTunnel = false
 	}
 
 	// For outbound case, we selectively add tunnel info if the other side supports the tunnel
-	if supportsTunnel {
+	if dir != model.TrafficDirectionInboundVIP && supportsTunnel {
+		// Support connecting to server side waypoint proxy, if the destination has one. This is for sidecars and ingress.
+		if dir == model.TrafficDirectionOutbound && !b.proxy.IsWaypointProxy() && !b.proxy.IsAmbient() {
+			workloads := findWaypoints(b.push, e)
+			if len(workloads) > 0 {
+				// TODO: load balance
+				tunnelAddress = workloads[0].String()
```

Comment:
Will this translate multi eps with same waytpoint address?

Reply:
Same waypoint yes, but the endpoints are not identical, so we cannot merge them and add `weight`. Each one has two addresses associated: the waypoint, and the destination (set in `:authority`) we want.